### ISSUE TITLE
Stack exhausted when using html2text on a very deeply-nested HTML document

### ISSUE THREAD

**Issue by djahandarie** [jugglerchris/rust-html2text#2]
It looks like the way `dom_to_render_tree` (and probably `render_tree_to_string` too, but it doesn't get that far) is written results in a super deep stack when you have a deep HTML tree. This could probably be avoided if the function were CPS-transformed, or perhaps some other solution.

**Comment by jugglerchris** [comment]
Hi!  Thanks for the report!

You're right, since it walks the tree by recursing there's the possibility of stack overflow.  Is this something you've come across in practice, or is it theoretical?  Do you have an example deeply nested file which you can share?  I haven't personally hit any which overflowed the stack, but my primary use case so far is HTML e-mail (which is probably relatively simple compared to real web pages).

**Comment by djahandarie** [comment]
Yeah, I ran into this in practice. I'm also using it for HTML emails actually!

This occurred when a buggy email client sent an HTML email where a bunch of the text wasn't escaped. And the body was a massive email thread, full of things like `From: "person a" <emaila@blah>, "person b" <emailb@blah>`. Of course, as HTML, each of those turns into an opening tag. When you have thousands of emails addresses, boom. And now that email is forever in my corpus, and does have valid information that needs to be parsed in it too, so I can't just discard it.

I unfortunately can't share the exact file, but if you just create a file full of `<x>` repeating over and over that should trigger the relevant case I think. All other HTML renderers I checked had no issue rendering it (certainly chrome, even things like mutt, etc.).

This probably isn't high priority for you, but I figured I'd at least record the issue! Thanks!

**Comment by jugglerchris** [comment]
I've committed a (failing, but ignored for now) unit test to make sure I remember.  I think it's important to fix, but I've got a few other things ahead of it in the queue.

My first thought would be to maintain an explicit stack of work in `dom_to_render_tree`, pushing context onto it instead of recursing into `children_to_render_nodes` and popping when the current set of sibling nodes is fully processed.
